hydromt/raster.py

```diff
@@ -262,53 +267,66 @@ def crs(self) -> CRS:
         """Return horizontal Coordinate Reference System."""
         # return horizontal crs by default to avoid errors downstream
         # with reproject / rasterize etc.
-        if "crs_wkt" not in self.attrs:
-            self.set_crs()
-        if "crs_wkt" in self.attrs:
-            crs = pyproj.CRS.from_user_input(self.attrs["crs_wkt"])
-            return crs
+        if self._crs is not None:
+            crs = self._crs
+        else:
+            crs = self.set_crs()
+        return crs
 
-    def set_crs(self, input_crs=None):
+    def set_crs(self, input_crs=None, write_crs=True) -> CRS:
         """Set the Coordinate Reference System.
 
         Arguments
         ---------
         input_crs: int, dict, or str, optional
             Coordinate Reference System. Accepts EPSG codes (int or str)
             and proj (str or dict)
+        write_crs: bool, optional
+            If True (default), write CRS to attributes.
         """
         crs_names = ["crs_wkt", "crs", "epsg"]
         names = list(self._obj.coords.keys())
         if isinstance(self._obj, xr.Dataset):
             names = names + list(self._obj.data_vars.keys())
         # user defined
-        if input_crs is not None:
+        if isinstance(input_crs, (int, str, dict)):
             input_crs = pyproj.CRS.from_user_input(input_crs)
         # look in grid_mapping and data variable attributes
+        elif input_crs is not None and not isinstance(input_crs, CRS):
+            raise ValueError(f"Invalid CRS type: {type(input_crs)}")
         else:
+            crs = None
             for name in crs_names:
-                # check default > GEO_MAP_COORDS attrs
-                crs = self._obj.coords[GEO_MAP_COORD].attrs.get(name, None)
-                if crs is None:  # global attrs
-                    crs = self._obj.attrs.pop(name, None)
-                for var in names:  # data var and coords attrs
-                    if name in self._obj[var].attrs:
-                        crs = self._obj[var].attrs.pop(name)
-                        break
-                if crs is not None:
-                    # avoid Warning 1: +init=epsg:XXXX syntax is deprecated
-                    crs = crs.removeprefix("+init=") if isinstance(crs, str) else crs
-                    try:
-                        input_crs = pyproj.CRS.from_user_input(crs)
-                        break
-                    except RuntimeError:
-                        pass
+                # check default > GEO_MAP_COORDS attrs, then global attrs
+                if name in self.attrs:
+                    crs = self.attrs.get(name)
+                    break
+                if name in self._obj.attrs:
+                    crs = self._obj.attrs.pop(name)
+                    break
+            if crs is None:  # check data var and coords attrs
+                for var in names:
+                    for name in crs_names:
```

Comment:
You could use `product` from itertools here..
That's just my taste. This is also fine